samples/net/http_server/src/main.c

```diff
@@ -167,6 +171,31 @@ static int http_response_soft_404(struct http_server_ctx *ctx)
 			     HTML_FOOTER);
 }
 
+static int http_response_auth(struct http_server_ctx *ctx)
+{
+	return http_response(ctx, HTTP_STATUS_200_OK, HTML_HEADER
+			     "<h2><center>user: "HTTP_AUTH_USERNAME
+			     ", password: "HTTP_AUTH_PASSWORD
+			     "</center></h2>" HTML_FOOTER);
+}
+
+int http_response_401(struct http_server_ctx *ctx)
+{
+	/* Wait 2 secs for the reply with proper credentials */
+	return http_response_wait(ctx, HTTP_401_STATUS_US, NULL, K_SECONDS(2));
+}
+
+static int http_basic_auth(struct http_server_ctx *ctx)
+{
+	const char *auth_str = "Authorization: Basic "HTTP_AUTH_CREDENTIALS;
+
+	if (strstr(ctx->req.field_values[0].key, auth_str)) {
```

Comment:
A couple of problems in this line: you're using `strstr()`, which depends on both strings being NUL-terminated, on a user-controlled value.  This is pretty dangerous.

Also, `strstr()` is used when you want to find a substring within a string, not compare two strings; although different, I can envision attacks similar to those you can pull off with a De Bruijn sequence.

Use `strncmp()` or `memcmp()` here, and use the length of `auth_str` instead, as that's something you control. If you declare `auth_str` as an array rather than a pointer, `sizeof()` can be used to calculate the length in compile time.